pkg/cvo/metrics.go

```diff
@@ -489,12 +489,16 @@ func (m *operatorMetrics) Collect(ch chan<- prometheus.Metric) {
 		if version == "" {
 			klog.V(2).Infof("ClusterOperator %s is not setting the 'operator' version", op.Name)
 		}
-		g := m.clusterOperatorUp.WithLabelValues(op.Name, version)
-		if resourcemerge.IsOperatorStatusConditionTrue(op.Status.Conditions, configv1.OperatorAvailable) {
-			g.Set(1)
-		} else {
-			g.Set(0)
+		isUp := 0
+		var reason := "NoAvailableCondition"
+		if condition := resourcemerge.FindOperatorStatusCondition(op.Status.Conditions, configv1.OperatorAvailable); condition != nil {
```

Comment:
Metric `cluster_operator_up` is equivalent to metric `cluster_operator_conditions{condition="Available"}`.
What is the point of keeping it?

Reply:
> Metric `cluster_operator_up` is equivalent to metric `cluster_operator_conditions{condition="Available"}`.

`cluster_operator_up` also includes the `operator` version from the ClusterOperator.  And it detects "ClusterOperator exists but `Available` is unset", where `cluster_operator_conditions` has no time-series.

Reply:
> Also semantic of `cluster_operator_conditions` metric is clear (I do not have to look at the code producing that metric to know what it does)...

I'm completely onboard with improving help-text for `cluster_operator_up`.

Reply:
I've expanded `cluster_operator_up`'s help text in 9f587ce9 -> edf2574d.

Reply:
> cluster_operator_up also includes the operator version from the ClusterOperator.
Couldn't the `cluster_operator_conditions` metric ship the version as well?

> And it detects "ClusterOperator exists but Available is unset", where cluster_operator_conditions has no time-series.
Well, here `cluster_operator_up` will be set to 0 whether the Available condition is unset to set to 0. Only `cluster_operator_conditions{condition="Available"}` metric can be use to distinguish it from being unset or set to 0.

> I'm completely onboard with improving help-text for cluster_operator_up.
Thanks for that, but as said, just with its name, I know what `cluster_operator_conditions` does. I have to dig further to know what `cluster_operator_up` does.

Last, please note that in some other request (OSD-8320), we have been asked to use the reason of the Degraded condition in conjunction with `cluster_operator_up` to inhibit `ClusterOperatorDown` alert.
Maybe this is a misstake as it would be strange to pick up the value from one metric (`cluster_operator_conditions{condition="Available"}`) and the reason from an other (`cluster_operator_conditions{condition="Degraded"}`); but hits is puzzling to me and need to be clarified.

Reply:
> Couldn't the `cluster_operator_conditions` metric ship the version as well?

It could, but the version is ClusterOperator-scoped, and not condition-scoped.  `cluster_operator_up` is ClusterOperator-scoped, so it's a good fit there.  `cluster_operator_conditions` is condition-scoped, so it's not a good fit there (would you copy the same version on to time-series for every one of the operator's conditions?).  You could obviously also add a new ClusterOperator-scoped metric whose sole purpose was exposing the version information, but :shrug:, that seems like a lot of churn, and I'm not clear on the benefit.

> Only `cluster_operator_conditions{condition="Available"}` metric can be use to distinguish it from being unset or set to 0.

With this pull request, you can make that distinction by seeing if the `cluster_operator_up == 0` `reason` is `NoAvailableCondition` or otherwise.  But as far as "when to page an admin?" goes, I don't think that distinction is all that important.  Both are bad enough that we want to be paging admins.

> I know what `cluster_operator_conditions` does. I have to dig further to know what `cluster_operator_up` does.

I don't mind if other cluster-version-operator approvers want to override me, but personally "many metrics consumers are able to intuit the semantics without having to read docs" is a useful property, and worthy of aspiring to, but not something that must be delivered in order for a metric to exist.

> Last, please note that in some other request (OSD-8320), we have been asked to use the reason of the Degraded condition in conjunction with `cluster_operator_up` to inhibit `ClusterOperatorDown` alert.

That seems like an anti-pattern to me, since `Degraded=True` is theoretically orthogonal to `Available=False`.  And it's `Available=False` (or absent) that are driving `ClusterOperatorDown`.  We've had it decoupled from `Degraded` since #550.  That said, my push-back is just that I don't want to re-mix `Degraded`ness into `ClusterOperatorDown`; I have no problem at all if consumers like OSD decide to join `ClusterOperatorDown` with `Degraded` information when deciding if/how to respond.

Reply:
Thanks for the feedback Trevor.
As my remark is more a matter of preference, I am happy to close this discussion.
My worries linked to OSD-8320 which was asking to use a different condition for the reason. From what I understand using the reason coming from the Available condition is perfectly fine with the silences defined there:
https://github.com/openshift/configure-alertmanager-operator/commit/db85a6dd2c55e26c0a2e8e381d999d4cc8eefbf6
(although I wonder if it would be better to not silence the alert and let CAD process it by sending an automated SL to the customer)